---
title: "TCGA Genomics Data In-Class Demo"
author: Mahija Mogalipuvvu, adapted from Nicole Black, Wade Boohar
date: 07/17/22
updated: 10/2/24
---

###### (1) Setting Up! 

```{r setup}
    knitr::opts_knit$set(root.dir = normalizePath("/home1/jlin7423/490_cluster/analysis_data")) 
```

```{r}
# install and load maftools, TCGAbiolinks, and ggplot2
if(!require("maftools")) install.packages("maftools")
if(!require("TCGAbiolinks")) install.packages("TCGAbiolinks")
if(!require("ggplot2")) install.packages("ggplot2")

library(maftools)
library(TCGAbiolinks)
library(ggplot2)
```

Before we get into MAF, we need to read in the clinical data (created in Introduction_to_Clinical.Rmd) so it can be merged with our MAF data later on.

```{r Exercise 1.1}
clinical <- read.csv("/home1/jlin7423/490_cluster/analysis_data/brca_clinical_data.csv")
```

Check that the bcr_patient barcode has been renamed to Tumor_Sample_Barcode which should have been done in the Introduction_to_Clinical.Rmd homework. This allows the MAF package to read our clinical file. If not, do the following commands to change the column names appropriately and then re-save the clinical data frame to your computer. 


```{r Exercise 1.2}
colnames(clinical)[ colnames(clinical) == "bcr_patient_barcode" ] <- "Tumor_Sample_Barcode"

# change the file path! make sure it's in your  folder
write.csv(clinical, "/home1/jlin7423/490_cluster/analysis_data/brca_clinical_data.csv", row.names = FALSE)
```

###### (2) MAF Package

Now, we can query in the MAF files using GDCquery()/GDCdownload()/GDCprepare() and read.maf(). 

```{r Exercise 2.1}
maf_query <- GDCquery(
    project = "TCGA-BRCA", 
    data.category = "Simple Nucleotide Variation", 
    access = "open", # we only have access to somatic mutations which are open access
    data.type = "Masked Somatic Mutation", 
    workflow.type = "Aliquot Ensemble Somatic Variant Merging and Masking"
)

#GDCdownload(maf_query)

maf <- GDCprepare(maf_query) # as long as it runs, ignore any errors

maf_object <- read.maf(maf = maf, 
                       clinicalData = clinic,
                       isTCGA = TRUE)
```

###### (3) Exploring the MAF Object

Call str() on maf_object.

```{r Exercise 3.1}
# write code here
```

Notice the @ and $ signs. What do each of them represent?

  @ represents:
  $ represents:
  
List all of the data frames within the maf_object by typing maf_object@ then either waiting for a popup, or if no popup appears, by hitting tab.

```{r Exercise 3.2}
# write code here
```

Call head() and str() on the @data df and the @clinical.data df. These will be the main data frames we use in our MAF analysis.

```{r Exercise 3.3}
# write code here
```

View the first 110 items in Tumor_Sample_Barcode from @data and @clinical.data.

```{r Exercise 3.4}
# write code here
```

What do you notice about these two lists?

View the Tumor_Sample_Barcode column and the HGVSp_Short column for all TP53 mutations (i.e. Hugo_Symbol == "TP53"). Hint: boolean masking!

```{r Exercise 3.5}
# write code here
```

Call head() on the Amino_acids column in @data.

```{r Exercise 3.6}
# write code here
```
Refer to a codon chart and amino acid abbreviation chart to answer the following:
 - What mutation could cause a S/R change?
 - What about an A/T change?

Call head() on the HGVSp_Short column in @data.

```{r Exercise 3.7}
# write code here
```

What does p.P728Afs*11 mean? (Do some google-ing!)

