---
title: 'SKCM Analysis Fall 2024 Review Project'
author: Joshua Lin, with assistance from Wade Boohar, Mahija Mogalipuvvu, and Ethen Chen
---

```{r}
# Installing + Loading BiocManager, maftools, TCGAbiolinks, ggplot2 libraries
if (!require("knitr", quietly = TRUE)) install.packages("knitr")

if (!require("BiocManager")) install.packages("BiocManager")
BiocManager::install(version = "3.19")

if (!require("TCGAbiolinks", quietly = TRUE)) BiocManager::install("TCGAbiolinks")

if(!require("ggplot2")) install.packages("ggplot2")

if (!require('survival')) install.packages("survival")

if (!require('survminer')) install.packages("survminer")

if (!require("DESeq2", quietly = TRUE)) BiocManager::install("DESeq2")

if (!require("EnhancedVolcano", quietly = TRUE)) BiocManager::install("EnhancedVolcano")

if (!require("SummarizedExperiment", quietly = TRUE)) BiocManager::install("SummarizedExperiment")

if (!require("sesameData", quietly = TRUE)) BiocManager::install("sesameData")

if (!require("sesame", quietly = TRUE)) BiocManager::install("sesame")

if (!require("limma", quietly = TRUE)) BiocManager::install("limma")

library(BiocManager)
library(TCGAbiolinks)
library(ggplot2)
library(survival)
library(survminer)
library(DESeq2)
library(EnhancedVolcano)
library(SummarizedExperiment)
library(sesame)
library(sesameData)
library(limma)
```

```{r setup}
   dir.create('/home1/jlin7423/490_cluster/r_review_joshlin/outputs')
  knitr::opts_knit$set(root.dir = normalizePath("/home1/jlin7423/490_cluster/r_review_joshlin/outputs"))
```

Query, Download, and Retrieve TCGA SKCM Data
SKCM Drug + Radiation Dataframe Retrieval
```{r}
clin_query <- GDCquery(project = "TCGA-SKCM",
                       data.category = "Clinical",
                       data.type = "Clinical Supplement",
                       data.format = 'BCR Biotab')
GDCdownload(clin_query)
clinical.BCRtab.all <- GDCprepare(clin_query)
clinic <- clinical.BCRtab.all$clinical_patient_skcm[-c(1,2),]
colnames(clinic)[ colnames(clinic) == "bcr_patient_barcode" ] <- "Tumor_Sample_Barcode"

# Radiation + Drug Dataframes:
rad <- clinical.BCRtab.all$clinical_radiation_skcm[-c(1,2),]
drug <- clinical.BCRtab.all$clinical_drug_skcm[-c(1,2),]

write.csv(clinic, "skcm_clinical_data.csv", row.names=FALSE)
write.csv(rad, "skcm_radiation_data.csv", row.names=FALSE)
write.csv(drug, "skcm_drug_data.csv", row.names=FALSE)
```

RNAclinical, RNAgenes, RNAcounts
```{r}
# Queue, Download, and Retrieve RNA_se
rna_query <- GDCquery(project ="TCGA-SKCM",
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification",
                      workflow.type = "STAR - Counts")
GDCdownload(rna_query)
rna_se <- GDCprepare(rna_query)

# rna_clinical
rna_clinical <- rna_se@colData
rna_clinical <- as.data.frame(rna_clinical)
column_mask <- ifelse(colnames(rna_clinical) %in% c('treatments', 'primary_site', 'disease_type'), F, T)
rna_clinical <- rna_clinical[ , column_mask]

# rna_genes
rna_genes <- rna_se@rowRanges@elementMetadata
rna_genes <- as.data.frame(rna_genes)

# rna_counts
rna_counts <- rna_se@assays@data$unstranded
rna_counts <- as.data.frame(rna_counts)

# Processing
rna_clinical$age_category <- rna_clinical[ifelse(rna_clinical$age_at_index <= 58, 'young', 'old'), ]
colnames(rna_counts) <- rna_clinical$patient
rownames(rna_counts) <- rna_genes$gene_id

# Writing CSVs
write.csv(rna_clinical, 'skcm_rna_clinical.csv')
write.csv(rna_genes,'skcm_rna_genes.csv')
write.csv(rna_counts, 'skcm_rna_counts.csv')
```

Read in clinic, rad, and drug dataframes (csv)
Read in rna_clinical, rna_genes, and rna_counts
```{r}
clinic <- read.csv("skcm_clinical_data.csv")
rad <- read.csv("skcm_radiation_data.csv")
drug <- read.csv("skcm_drug_data.csv")

rna_clinical <- read.csv("skcm_rna_clinical.csv")
rna_genes <- read.csv("skcm_rna_genes.csv")
rna_counts <- read.csv("skcm_rna_counts.csv", row.names=1)
```

KM Plot (also DESeq2) Pre-Processing
```{r}
# RNA Clinical Definition (Metastatic + Primary solid Tumor)
rna_clinical$definition <- factor(rna_clinical$definition)

# Cleaning NAs - ensuring only 
rm_na_mask <- !is.na(rna_clinical$definition)
rna_counts <- rna_counts[ ,rm_na_mask]
rna_clinical <- rna_clinical[rm_na_mask, ]

# Cleaning out Solid Tissue Normal + Additional Metastatic -> only Metastatic and Primary solid Tumor
rm_norm_mask <- rna_clinical$definition != "Solid Tissue Normal" & rna_clinical$definition != 'Additional Metastatic'
rna_counts <- rna_counts[ ,rm_norm_mask]
rna_clinical <- rna_clinical[rm_norm_mask, ]

# Creating clean_rna_clinical for KM Plot
clean_rna_clinical <- rna_clinical

# Survival Time (# Days survived after diagnosis)
clean_rna_clinical$survival_time <- as.numeric(ifelse(clean_rna_clinical$vital_status == 'Dead', clean_rna_clinical$days_to_death, clean_rna_clinical$days_to_last_follow_up))
clean_rna_clinical[!is.na(clean_rna_clinical$survival_time),]

# Death Event
clean_rna_clinical$death_event <- clean_rna_clinical$vital_status == "Dead"
```

Survminer Object
```{r}
survival_object <- Surv(time = clean_rna_clinical$survival_time,
                        event = clean_rna_clinical$death_event)

fit_object <- survfit(survival_object ~ definition, data = clean_rna_clinical)
```

KM Plotting
```{r}
survplot <- ggsurvplot(fit_object,
                       pval=TRUE,
                       ggtheme = theme(plot.margin = unit(c(1,1,1,1), "cm")),
                       legend = 'right')

KM_plot <- survplot$plot + theme_bw() + theme(axis.title = element_text(size=20), 
                                              axis.text = element_text(size=16),
                                              legend.title = element_text(size=14),
                                              legend.text = element_text(size=12))

KM_plot
```

DESEq2 Pre-Processing
```{r}
# Treatment (One-Hot Encoding) Columns + Factoring
rna_clinical$Radiation <- factor(ifelse(rna_clinical$patient %in% rad$bcr_patient_barcode, 1, 0))
rna_clinical$Chemo <- factor(ifelse(rna_clinical$patient %in% drug$bcr_patient_barcode & drug$pharmaceutical_therapy_type == "Chemotherapy", 1, 0))
rna_clinical$Immuno <- factor(ifelse(rna_clinical$patient %in% drug$bcr_patient_barcode & drug$pharmaceutical_therapy_type == "Immunotherapy", 1, 0))
rna_clinical$Molecular <- factor(ifelse(rna_clinical$patient %in% drug$bcr_patient_barcode & drug$pharmaceutical_therapy_type == "Targeted Molecular therapy", 1, 0))
rna_clinical$Vaccine <- factor(ifelse(rna_clinical$patient %in% drug$bcr_patient_barcode & drug$pharmaceutical_therapy_type == "Vaccine", 1, 0))

# Factoring covariates + removing NAs
rna_clinical$race <- factor(rna_clinical$race)
rna_clinical$gender <- factor(rna_clinical$gender)
rna_clinical$vital_status <- factor(rna_clinical$vital_status)

# Race NAs
rm_na_race <- !is.na(rna_clinical$race)
rna_counts <- rna_counts[ ,rm_na_race]
rna_clinical <- rna_clinical[rm_na_race, ]

# Gender NAs
rm_na_gender <- !is.na(rna_clinical$gender)
rna_counts <- rna_counts[ ,rm_na_gender]
rna_clinical <- rna_clinical[rm_na_gender, ]

# Vital Status NAs
rm_na_vital <- !is.na(rna_clinical$vital_status)
rna_counts <- rna_counts[ ,rm_na_vital]
rna_clinical <- rna_clinical[rm_na_vital, ]

# Filtering genes that have total expression < 20
keep_genes <- rowSums(rna_counts) >= 20
filtered_rna_counts <- rna_counts[keep_genes, ]
```

DESeq2 Analysis
```{r}
dds <- DESeqDataSetFromMatrix(countData = rna_counts,
                              colData = rna_clinical,
                              design= ~race + gender + vital_status + Radiation + Chemo + Immuno + Molecular + Vaccine + definition)

dds_obj <- DESeq(dds)

resultsNames(dds_obj)

results <- results(dds_obj, format = "DataFrame", contrast = c("definition", "Metastatic", "Primary solid Tumor"))

results <- data.frame(results)

# Save Results.csv for Methylation analysis  
write.csv(results, "skcm_DESEq2_results.csv", row.names=FALSE)
```

Enhanced Volcano
```{r}
results <- read.csv('skcm_DESEq2_results.csv')
results$gene_name <- rna_genes$gene_name
results$'-log10(padj)' <- -log(results$padj, 10)

EnhancedVolcano(results,
                lab = results$gene_name,
                x = 'log2FoldChange',
                y = '-log10(padj)',
                title = 'Sample Definition: Metastatic vs Primary solid Tumor',
                pointSize = 1.0,
                labSize = 5.0)
```

Retrieving Methylation Data:
```{r}
query <- GDCquery(project = "TCGA-SKCM",
                  data.category = "DNA Methylation",
                  data.type = "Methylation Beta Value",
                  platform = "Illumina Human Methylation 450",
                  data.format = "Dataframe")
GDCdownload(query)
methylation450 <- GDCprepare(query)
```

```{r}
# Obtaining respective dataframes
methylation_clinical <- as.data.frame(methylation450@colData)
betas <- as.data.frame(methylation450@assays@data@listData)
cpg_sites <- as.data.frame(methylation450@rowRanges@elementMetadata)

# Methylation Clinical Preprocessing + Saving as CSV
column_mask <- ifelse(colnames(methylation_clinical) %in% c('treatments', 'primary_site', 'disease_type'), F, T)
methylation_clinical <- methylation_clinical[,column_mask]
write.csv(methylation_clinical, 'skcm_methylation_clinical.csv')

# Filtering CPG sites (w/ multiple -,; NA, CPG row NA)
site_mask <- !grepl('-', cpg_sites$gene) & !grepl(';', cpg_sites$gene) & !is.na(cpg_sites$gene) & complete.cases(betas)
betas <- betas[site_mask,]
cpg_sites <- cpg_sites[site_mask,]

# Saving to CSV
write.csv(betas, 'skcm_methylation_betas.csv')
write.csv(cpg_sites, 'skcm_cpg_sites.csv')
```

Reading CSVs for methylation
```{r}
methylation_clinical <- read.csv('skcm_methylation_clinical.csv')
betas <- read.csv('skcm_methylation_betas.csv', row.names = 1)
cpg_sites <- read.csv('skcm_cpg_sites.csv')
```

Differential Methylation
```{r}
# Fitting linear values - Target column = tumor - remove normal tissue
normal_tissue_mask <- methylation_clinical$definition == "Solid Tissue Normal"
methylation_clinical$tumor <- !normal_tissue_mask

# Transforming m values
mval <- t(apply(betas, 1, function(x) log2(x/(1-x))))

# Depends on tumor, linear fitting + interpolation 
design <- model.matrix(~ tumor, data = methylation_clinical)
fit <- lmFit(mval, design)
fit2 <- eBayes(fit)
```

Differential Methylation Volcano Plot (ggplot)
Blue - Right = hypermethylated in tumor (compared to normal baseline)
Blue - Left = undermethylated in tumor (compared to normal baseline)
```{r}
dat <- data.frame(foldchange = fit[["coefficients"]][,2], logPvalue =  -log10(p.adjust(fit2[["p.value"]][,2],method='BY')), geneName = cpg_sites$gene)
dat$threshold <- as.factor(abs(dat$foldchange) < 1)

#Visualization
cols <- c("TRUE" = "grey", "FALSE" = "blue")
ggplot(data=dat, aes(x=foldchange, y = logPvalue, color=threshold)) +
  geom_point(alpha=.2, size=0.6) +
  scale_colour_manual(values = cols) +
  geom_vline(xintercept = 1, colour="#990000", linetype="dashed") + 
  geom_vline(xintercept = - 1, colour="#990000", linetype="dashed") +
  geom_hline(yintercept = 2, colour = "#990000", linetype="dashed") +
  theme(legend.position="none") +
  xlab("Fold Change") +
  ylab("-log10 p value") +
  theme_bw() +
  theme(legend.position = "none")
```

Load in results, rna_counts, rna_genes, and rna_counts saved earlier
```{r}
results <- read.csv("skcm_DESEq2_results.csv")
results$gene_name <- rna_genes$gene_name
results$'-log10(padj)' <- -log(results$padj, 10)

rna_clinical <- read.csv("skcm_rna_clinical.csv")
rna_genes <- read.csv("skcm_rna_genes.csv")
rna_counts <- read.csv("skcm_rna_counts.csv", row.names=1)

```

Looking at downregulated and Hypermethylated genes:
```{r}
downregulated <- results[(results$log2FoldChange < -3), 'gene_name']
hypermethylated <- dat[dat$foldchange > 2, 'geneName']
interest_genes <- intersect(downregulated, hypermethylated)
```

```{r}
GENE <- "SCNN1A"

gene_counts_mask <- rna_genes$gene_name == GENE
gene_betas_mask <- cpg_sites$gene == GENE

rna_clinical_tumor <- rna_clinical$definition == "Primary solid Tumor"
methylation_clinical_tumor <- methylation_clinical$definition == "Primary solid Tumor"

rna_clinical_metastatic <- rna_clinical$definition == "Metastatic"
methylation_clinical_metastatic <- methylation_clinical$definition == "Metastatic"

rna_tumor <- as.numeric(rna_counts[gene_counts_mask, rna_clinical_tumor])
methylation_tumor <- betas[gene_betas_mask, methylation_clinical_tumor]

rna_metastatic <- as.numeric(rna_counts[gene_counts_mask, rna_clinical_metastatic])
methylation_metastatic <- betas[gene_betas_mask, methylation_clinical_metastatic]
```

Boxplot to see transcriptomic differences between Metastatic and Tumor groups
```{r}
boxplot(rna_metastatic, rna_tumor, xlab='Group', ylab='Counts', names=c('Metastatic', 'Tumor'))
```

Barplot for the GENE to see methylation differences between the two groups
```{r}
plot(1:dim(methylation_metastatic)[1], apply(methylation_metastatic, 1, mean), col='blue', type='h', lwd=3, ylab='Average Beta', xlab='CpG sites')
lines(1.3:(dim(methylation_metastatic)[1]+0.3), apply(methylation_tumor, 1, mean), col='red', type='h', lwd=3)
```


